compiler/src/abi/builder.rs

```diff
@@ -0,0 +1,281 @@
+use crate::abi::elements::{
+    Contract, Event, EventField, FuncInput, FuncOutput, Function, FunctionType, ModuleABIs,
+    VariableType,
+};
+use crate::errors::CompileError;
+use std::collections::HashMap;
+use vyper_parser::ast as vyp;
+
+type TypeDefs<'a> = HashMap<&'a str, &'a vyp::TypeDesc<'a>>;
+type TypeDef<'a> = (&'a str, &'a vyp::TypeDesc<'a>);
+
+/// Parse a map of contract ABIs from the input `module`.
+pub fn module<'a>(module: &'a vyp::Module<'a>) -> Result<ModuleABIs, CompileError> {
+    let type_defs = module
+        .body
+        .iter()
+        .filter(|s| is_type_def(&s.node))
+        .map(|s| type_def(&s.node))
+        .collect::<Result<TypeDefs<'a>, _>>()?;
+
+    Ok(ModuleABIs(
+        module
+            .body
+            .iter()
+            .filter(|s| is_contract_def(&s.node))
+            .map(|s| contract_def(&type_defs, &s.node))
+            .collect::<Result<HashMap<_, _>, _>>()?,
+    ))
```

Comment:
we don't need to do this here but i'll call your attention to `scan`: https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.scan

you could collect the type defs into the iterator state which is supplied to every call to `contract_def` and then we can skip the double iteration over the module.... would require some special handling for forward references but i don't know if we are treating those in a special way yet or not